Cilium docs Dark Mode Support

UtkarshSiddhpura opened the pull request:
# Add Dark Mode Support (Resolves [cilium/issues/31391](https://github.com/cilium/cilium/issues/31391))

Integrated dark mode functionality from [MrDogeBro/sphinx_rtd_dark_mode](https://github.com/MrDogeBro/sphinx_rtd_dark_mode) into the Cilium Sphinx theme. Reused core CSS and JS for theme toggling, with modifications to support Pygments monokai theme, sphinx-tabs dark styling, DocSearch v4 dark mode, and Cilium-specific components. Attribution is provided at the top of each relevant file.

![Cilium-Dark-Theme](https://github.com/user-attachments/assets/0930d5a4-7110-402b-90d1-a39e7081891d)

## Changes
- Added pygments_dark_style with monokai theme as i find it to be the most complete for dark theme styles.
- Updated docsearch from v2 (no dark theme support) to v4 which has full dark-theme support.
- Updated to support sphinx-tabs (code tabs) dark theme with our theme-switcher setup.
- Used dark variant of Cilium logo for dark theme.
- Default theme as light.


## Testing this Update
1. Update [Documentation/requirements.txt#L6](https://github.com/cilium/cilium/blob/main/Documentation/requirements.txt#L6) to use my forked theme latest commit hash:

```
sphinx-rtd-theme-cilium @ git+https://github.com/UtkarshSiddhpura/sphinx_rtd_theme.git@a040208de143286b4198c51180a1e83611448f32
```
2. **Build the docs site**:
```bash
make html
```
3. run `Documentation/_build/html` website with any http server

Made it modular with separate dark mode files and minimal config changes, so any future updates or conflicts (very unlikely) can be handled easily. If any issues come up, I’ll be around to help.

## Issues
- Architecture diagrams and component images have white backgrounds that don't adapt to dark mode
<img width="1377" height="885" alt="image" src="https://github.com/user-attachments/assets/8f743fea-b0ce-402e-b327-e4bbfe52225e" />


## References
- [sphinx_rtd_dark_mode](https://github.com/MrDogeBro/sphinx_rtd_dark_mode)
- [Are there any other themes besides the default Light mode?](https://support.algolia.com/hc/en-us/articles/13924080358545-Are-there-any-other-themes-besides-the-default-Light-mode)
- [sphinx-tabs dark mode issue](https://github.com/executablebooks/sphinx-tabs/issues/152)

sandipkumardey commented:
Just checked the PR. Its good to merged. Maybe we need a few more customizations from the UI point of view and color, but overall, good work 👍

qmonnet commented:
Netlify preview (from [this PR](https://github.com/cilium/cilium/pull/40802)): https://deploy-preview-40802--docs-cilium-io.netlify.app/

Cc @xmulligan

qmonnet commented:
I think it looks great! It's absolutely fine if some pictures have their white background, we'll fix them on a case-by-case basis if we really want a uniform background.

> - Added pygments_dark_style with monokai theme as i find it to be the most complete for dark theme styles.
> - Updated docsearch from v2 (no dark theme support) to v4 which has full dark-theme support.
> - Updated to support sphinx-tabs (code tabs) dark theme with our theme-switcher setup.
> - Used dark variant of Cilium logo for dark theme.

Any chance we could have the different items above as separate commits, please? It would help for maintenance, if we ever need a revert, or more likely if we need to address conflicts when rebasing on the upstream version of the theme.

> - Default theme as dark.

Could we keep the light theme as default for now? I'd rather have the dark theme as optional so we can run and fix/polish a little bit if required, and also because cilium.io doesn't have a dark mode yet so to keep consistent between the two.

qmonnet commented:
Also, one question: Where is the CSS for the dark theme coming from, did you write it yourself? If not (if you imported portions of it from the dark-mode add-on for RTD, for example), should we provide attribution?

sandipkumardey commented:
I was thinking the same @qmonnet, let's keep the default theme as light only. Whatever minor fixes are needed, we can do them in future commits. And for the cilium.io website, I and Priya are working.

UtkarshSiddhpura commented:
> Also, one question: Where is the CSS for the dark theme coming from, did you write it yourself? If not (if you imported portions of it from the dark-mode add-on for RTD, for example), should we provide attribution?

Attribution is provided in the footer:
<img width="903" height="100" alt="image" src="https://github.com/user-attachments/assets/05aace5f-3db9-4242-8544-98af2a2ba699" />

Yes, it's mostly based on [MrDogeBro/sphinx_rtd_dark_mode](https://github.com/MrDogeBro/sphinx_rtd_dark_mode) for the basic dark mode styles, but I’ve also added custom styles and simplified it to better fit our theme fork..

qmonnet commented:
> Yes, it's mostly based on [MrDogeBro/sphinx_rtd_dark_mode](https://github.com/MrDogeBro/sphinx_rtd_dark_mode) for the basic dark mode styles

OK, if this means some of the CSS rules were copied, please make it very explicit (in your commit description, for example) what portions come from the add-on. I'd even add a comment at the top of the relevant files, in that case.

Attribution in the footer is a good idea, thanks for that

UtkarshSiddhpura commented:
> OK, if this means some of the CSS rules were copied, please make it very explicit (in your commit description, for example) what portions come from the add-on. I'd even add a comment at the top of the relevant files, in that case.

Will update in the commit as well as in the files 👍

UtkarshSiddhpura committed:
docs(theme): add dark mode base, modified from sphinx_rtd_dark_mode plugin

Signed-off-by: UtkarshSiddhpura <utkarshsiddhpura09876@gmail.com>

UtkarshSiddhpura committed:
docs(docsearch): upgrade docsearch to v4 to support dark theme

Signed-off-by: UtkarshSiddhpura <utkarshsiddhpura09876@gmail.com>

UtkarshSiddhpura committed:
docs(pygments): add monokai style for dark theme

Signed-off-by: UtkarshSiddhpura <utkarshsiddhpura09876@gmail.com>

UtkarshSiddhpura commented:
>Any chance we could have the different items above as separate commits, please? It would help for maintenance, if we ever need a revert, or more likely if we need to address conflicts when rebasing on the upstream version of the theme.

>Could we keep the light theme as default for now? I'd rather have the dark theme as optional so we can run and fix/polish a little bit if required, and also because cilium.io doesn't have a dark mode yet so to keep consistent between the two.

@qmonnet Done ✔️

qmonnet commented on sphinx_rtd_theme/__init__.py:
```diff
@@ -66,8 +66,14 @@ def setup(app):
     app.connect('config-inited', config_initiated)
-
+    # Add the dark mode configuration option
+    app.add_config_value("default_dark_mode", True, "html")
```
What does this option do? I don't see it used anywhere (but I may have missed it, I'm not super familiar with the Sphinx API).

qmonnet replied:
Apparently unused indeed, given that the last update removed it :slightly_smiling_face:

qmonnet reviewed (commented):
Thanks a lot for splitting the PR into several commits! Much better this way, in my opinion.

It looks good to me, although I've not dived into the JS/CSS details. If anyone (is better than me at JS/CSS and) has the time to look into it, in-depth reviews are welcome 😇. I just have one question, inline below.

I've updated the preview, still at https://deploy-preview-40802--docs-cilium-io.netlify.app/.

Can we have someone from @cilium/community take a look as well, please? Cc also @joestringer regarding the “docsearch” update, mostly FYI (search being the part I've always been least familiar with for our docs).

qmonnet reviewed (approved):
Looks good to me, thank you!

joestringer commented on sphinx_rtd_theme/static/js/docsearch.min.js:
Maybe you could share the process you used to update this dependency so we can follow along?

joestringer replied:
For a bit more context if other reviewers are looking over this: Basically this is a blob of code being replaced so I'd want to verify the origination of the library to ensure there's nothing unexpected here, just as a general sanity check. I also note that v4 is currently beta. I don't really understand what that means in terms of impact on how the theme works but we should be prepared to deal with any consequences that arise from that.

UtkarshSiddhpura replied:
@joestringer 
It is from the official docs here:
https://docsearch.algolia.com/docs/docsearch#without-package-manager

I copied the code directly from the jsdelivr CDN into our local files so we're not dependent on external links, even if the CDN goes down or changes. (Just like it was for docsearch v2)

joestringer replied:
Hmm. If I do `curl -L https://cdn.jsdelivr.net/npm/@docsearch/js@beta -o sphinx_rtd_theme/static/js/docsearch.min.js` with this PR checked out, it creates a diff. Maybe the upstream version has already moved on to a newer version?

UtkarshSiddhpura replied:
<img width="1532" height="114" alt="image" src="https://github.com/user-attachments/assets/a20d05ca-d0cd-4abf-9e35-23c4e90215b4" />
@joestringer Oo sorry my bad.. this is the only thing that differs.
The source map reference is included in the jsdelivr CDN blob because it's part of the original library build. However, we typically don't need debugging capabilities for the library's internal code, and the .map file often isn't available anyway. Causing unnecessary 404 errors. That's why I removed it.

UtkarshSiddhpura replied:
But We can absolutely use the CDN directly if you want, or use the version with the map reference with no problems.

UtkarshSiddhpura replied:
>I also note that v4 is currently beta. I don't really understand what that means in terms of impact on how the theme works but we should be prepared to deal with any consequences that arise from that.

I think docsearch changes (to v4) won't cause future problems because we're only using core, stable functionality, not the new experimental features like AskAI integration, BYO-LLM support, etc.. which maintains full backward compatibility with v3 stable API.

But if we need AI-powered search or advanced features later, we can easily add them by just configuring the askAi parameter.

joestringer replied:
I see a lot more diff if I just check out the PR and run the curl command above. The `sourceMappingURL` part seems innocent enough but I couldn't vouch for the changes to the blob of Javascript.

> I think docsearch changes (to v4) won't cause future problems because we're only using core, stable functionality, not the new experimental features

:ok_hand: I guessed this would be the case.

UtkarshSiddhpura replied:
> I see a lot more diff if I just check out the PR and run the curl command above. The `sourceMappingURL` part seems innocent enough but I couldn't vouch for the changes to the blob of Javascript.

I don't see any diff (except `sourceMappingURL`) btw `docsearch.min.js` present in this PR and cdn url (https://cdn.jsdelivr.net/npm/@docsearch/js@beta) on my side.

joestringer replied:
```
$ gh co --force UtkarshSiddhpura:cilium-docs-dark-mode
From github.com:cilium/sphinx_rtd_theme
 * branch                      refs/pull/26/head -> FETCH_HEAD
HEAD is now at 40a303947597 docs(pygments): add monokai style for dark theme
$ curl -L https://cdn.jsdelivr.net/npm/@docsearch/js@beta -o sphinx_rtd_theme/static/js/docsearch.min.js
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  302k  100  302k    0     0   312k      0 --:--:-- --:--:-- --:--:--  312k
$ git diff --stat
 sphinx_rtd_theme/static/js/docsearch.min.js | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)
$ git diff | head -c 1000
diff --git a/sphinx_rtd_theme/static/js/docsearch.min.js b/sphinx_rtd_theme/static/js/docsearch.min.js
index 32bd62d5d2fa..e77ef5048c1b 100644
--- a/sphinx_rtd_theme/static/js/docsearch.min.js
+++ b/sphinx_rtd_theme/static/js/docsearch.min.js
@@ -5,4 +5,5 @@
  * Do NOT use SRI with dynamically generated files! More information: https://www.jsdelivr.com/using-sri-with-dynamic-files
  */
 /*! @docsearch/js 4.0.0-beta.5 | MIT License | © Algolia, Inc. and contributors | https://docsearch.algolia.com */
-!function(e,t){"object"==typeof exports&&"undefined"!=typeof module?module.exports=t():"function"==typeof define&&define.amd?define(t):(e="undefined"!=typeof globalThis?globalThis:e||self).docsearch=t()}(this,(function(){"use strict";function e(e,t){(null==t||t>e.length)&&(t=e.length);for(var n=0,r=Array(t);n<t;n++)r[n]=e[n];return r}function t(e,t,n,r,u,o,a){try{var i=e[o](a),c=i.value}catch(e){return void n(e)}i.done?t(c):Promise.resolve(c).then(r,u)}function n(e){return function(){var 
```

joestringer replied:
Let me know if I'm doing something naive here, I'm actually not familiar with this code / repo at all, I just got pulled in via comment from Quentin above.

UtkarshSiddhpura replied:
Use this @joestringer 
```
curl -s https://cdn.jsdelivr.net/npm/@docsearch/js@beta | \
 diff -w - sphinx_rtd_theme/static/js/docsearch.min.js
 ```
 The -w flag (ignore whitespace)

joestringer replied:
I didn't realize there was any whitespace left after the JSON was squashed into that giant blob :sweat_smile: Thanks for the pointer, that resolves the diff.

joestringer replied:
I'll resolve this conversation as I think there are no other open discussion threads here.

qmonnet replied:
I never realised `diff` and `git diff` had an option to ignore white spaces (I knew the GitHub interfaces has it, though). Thanks for that, Utkarsh! :slightly_smiling_face:

joestringer reviewed (commented):
> Can we have someone from @cilium/community take a look as well

We're just in the process of setting up a cilium.io review group which I think will be better suited, but a few of the folks overlap with this team as well. I'll CC @xmulligan just for awareness as a backup notification.

qmonnet commented:
Nobody from @cilium/community seems to have an objection :cricket:. I'll merge this PR now, it won't come into effect until we've made the docs point to it in Cilium's repository, anyway.

Thanks a lot for your work!
